File: frontend/src/services/research-group.service.ts
import http from '@/lib/http'
import type { ApiResponse, PageResponse } from '@/schemas/base.schema'
import type {
  CreateResearchGroupRequest,
  MemberInfoResponse,
  MyResearchGroupResponse,
  ResearchGroupResponse,
  SecureResearchGroupResponse,
  UpdateResearchGroupRequest
} from '@/schemas/research-group.schema'
import type { UserResponse } from '@/schemas/user.schema'

export const researchGroupService = {
  filterGroups: (params: { page: number; limit: number; keyword?: string; type?: string; leaderId?: number }) =>
    http.get<ApiResponse<PageResponse<ResearchGroupResponse[]>>>('/research-groups', {
      params
    }),

  filterLeaders: (params: { keyword?: string }) =>
    http.get<ApiResponse<PageResponse<UserResponse[]>>>('/research-groups/leaders', {
      params
    }),

  filterMyGroups: (params: {
    page: number
    limit: number
    keyword?: string
    type?: string
    leaderId?: number
    isPrivate?: boolean
  }) =>
    http.get<ApiResponse<PageResponse<ResearchGroupResponse[]>>>('/research-groups/my', {
      params
    }),

  filterMyLeaders: (params: { keyword?: string }) =>
    http.get<ApiResponse<PageResponse<UserResponse[]>>>('/research-groups/my/leaders', {
      params
    }),

  filterOtherGroups: (params: { page: number; limit: number; keyword?: string; type?: string; leaderId?: number }) =>
    http.get<ApiResponse<PageResponse<ResearchGroupResponse[]>>>('/research-groups/other', {
      params
    }),

  filterOtherLeaders: (params: { keyword?: string }) =>
    http.get<ApiResponse<PageResponse<UserResponse[]>>>('/research-groups/other/leaders', {
      params
    }),

  findPublicGroupById: (id: number) => http.get<ApiResponse<ResearchGroupResponse>>(`/research-groups/${id}`),

  findGroupById: (id: number) => http.get<ApiResponse<ResearchGroupResponse>>(`/research-groups/${id}/find`),

  findMyGroupDetail: (id: number) => http.get<ApiResponse<MyResearchGroupResponse>>(`/research-groups/${id}/my-detail`),

  addMembers: (groupId: number, members: { username: string; role: string }[]) =>
    http.post<ApiResponse<ResearchGroupResponse>>(`/research-groups/${groupId}/members`, { members }),

  findMembersByGroupId: (id: number, params?: { page?: number; limit?: number }) =>
    http.get<ApiResponse<PageResponse<MemberInfoResponse[]>>>(`/research-groups/${id}/members`, { params }),

  filterManagedResearchGroups: (params: {
    page: number
    limit: number
    keyword?: string
    groupType?: string
    isPrivate?: boolean
    active?: boolean
  }) =>
    http.get<ApiResponse<PageResponse<SecureResearchGroupResponse[]>>>('/research-groups/managed', {
      params
    }),

  toggleStatus: (id: number) => http.patch<ApiResponse<SecureResearchGroupResponse>>(`/research-groups/${id}/status`),
  createGroup: (data: CreateResearchGroupRequest) =>
    http.post<ApiResponse<SecureResearchGroupResponse>>('/research-groups', data),
  updateGroup: (id: number, data: UpdateResearchGroupRequest) =>
    http.patch<ApiResponse<SecureResearchGroupResponse>>(`/research-groups/${id}`, data),

  filterUsersToInvite: (id: number, params: { page?: number; size?: number; keyword?: string }) =>
    http.get<ApiResponse<PageResponse<UserResponse[]>>>(`/research-groups/${id}/search-to-invite`, {
      params
    }),

  updateMemberRole: (groupId: number, data: { username: string; role: string }) =>
    http.patch<ApiResponse<string>>(`/research-groups/${groupId}/members/role`, data),
  findMyJoinedGroups: (params?: { page?: number; limit?: number }) =>
    http.get<ApiResponse<PageResponse<ResearchGroupResponse[]>>>('/research-groups/joined', { params }),
  filterResearchGroupsAdmin: (params: {
    page: number
    limit: number
    keyword?: string
    type?: string
    active?: boolean
    isPrivate?: boolean
  }) =>
    http.get<ApiResponse<PageResponse<SecureResearchGroupResponse[]>>>('/research-groups/admin', {
      params
    }),
  findGroupByIdAdmin: (id: number) => http.get<ApiResponse<SecureResearchGroupResponse>>(`/research-groups/admin/${id}`)
}

File: backend/src/main/java/iuh/labbooking/service/researchgroup/ResearchGroupService.java
package iuh.labbooking.service.researchgroup;

import iuh.labbooking.dto.request.researchgroup.AddMembersRequest;
import iuh.labbooking.dto.request.researchgroup.CreateResearchGroupRequest;
import iuh.labbooking.dto.request.researchgroup.UpdateResearchGroupRequest;
import iuh.labbooking.dto.request.researchgroup.UpdateMemberRoleRequest;
import iuh.labbooking.dto.response.base.PageResponse;
import iuh.labbooking.dto.response.researchgroup.MemberInfoResponse;
import iuh.labbooking.dto.response.researchgroup.MyResearchGroupResponse;
import iuh.labbooking.dto.response.researchgroup.ResearchGroupResponse;
import iuh.labbooking.dto.response.researchgroup.SecureResearchGroupResponse;
import iuh.labbooking.dto.response.user.UserResponse;
import iuh.labbooking.enums.GroupType;

import iuh.labbooking.model.ResearchGroup;

import java.util.List;
import java.util.Set;

public interface ResearchGroupService {
    Set<ResearchGroup> findEntitiesByIds(Set<Long> ids);

    SecureResearchGroupResponse createGroup(CreateResearchGroupRequest request);

    SecureResearchGroupResponse updateGroup(Long researchGroupId, UpdateResearchGroupRequest request);

    ResearchGroupResponse addMembers(Long researchGroupId, AddMembersRequest request);

    void updateMemberRole(Long researchGroupId, UpdateMemberRoleRequest request);

        PageResponse<List<ResearchGroupResponse>> filterGroups(int page, int limit, String keyword,
                        GroupType type, Long leaderId);

        PageResponse<List<UserResponse>> filterLeaders(String keyword);

        PageResponse<List<ResearchGroupResponse>> filterMyGroups(int page, int limit, String keyword,
                        GroupType type, Long leaderId, Boolean isPrivate);

        PageResponse<List<UserResponse>> filterMyLeaders(String keyword);

        PageResponse<List<ResearchGroupResponse>> filterOtherGroups(int page, int limit, String keyword,
                        GroupType type, Long leaderId);

        PageResponse<List<UserResponse>> filterOtherLeaders(String keyword);

    ResearchGroupResponse findPublicGroupById(Long researchGroupId);

    ResearchGroupResponse findGroupById(Long researchGroupId);

    MyResearchGroupResponse findMyGroupDetail(Long researchGroupId);

    PageResponse<List<MemberInfoResponse>> findMembersByGroupId(Long groupId, int page, int limit);

    PageResponse<List<SecureResearchGroupResponse>> filterManagedResearchGroups(int page, int limit, String keyword,
                                                                                GroupType type, Boolean isPrivate, Boolean active);

    SecureResearchGroupResponse toggleActiveStatus(Long researchGroupId);

    PageResponse<List<UserResponse>> filterUsersToInvite(int page, int size, Long groupId, String keyword);

    PageResponse<List<ResearchGroupResponse>> findMyJoinedGroups(int page, int limit);

    PageResponse<List<SecureResearchGroupResponse>> filterResearchGroupsAdmin(
            int page, int limit, String keyword, GroupType type, Boolean active, Boolean isPrivate);

    SecureResearchGroupResponse findGroupByIdAdmin(Long id);
}
